gemfire-core/src/main/java/com/gemstone/gemfire/cache/GemFireSparkConnectorCacheFactory.java

```diff
@@ -1,46 +1,157 @@
+        pf.create(gridName);
+      }
+      return cache;
+    }
+  }
+
+
+  private PoolFactory createAndConfigurePoolFactory(String remoteLocators) {
+    PoolFactory pf = PoolManager.createFactory();
+    pf.setReadTimeout(30000);
+
+    StringTokenizer remoteLocatorsTokenizer = new StringTokenizer(remoteLocators, ",");
+    DistributionLocatorId[] locators = new DistributionLocatorId[remoteLocatorsTokenizer
+        .countTokens()];
+    int i = 0;
+    while (remoteLocatorsTokenizer.hasMoreTokens()) {
+      locators[i++] = new DistributionLocatorId(remoteLocatorsTokenizer.nextToken());
+    }
+    List<ServerLocation> servers = new ArrayList<ServerLocation>();
+    for (DistributionLocatorId locator : locators) {
+      try {
+
+        InetSocketAddress addr = new InetSocketAddress(locator.getHost(), locator.getPort());
+        GetAllServersRequest req = new GetAllServersRequest("");
+        Object res = TcpClient.requestToServer(addr.getAddress(), addr.getPort(), req, 2000);
+        if (res != null) {
+          servers.addAll((List<ServerLocation>)((GetAllServersResponse)res).getServers());
+        }
+      } catch (Exception e) {
+        System.out.println("Unable to get remote gfe servers from locator = " + locator);
+      }
+    }
+
+    List<ServerLocation> prefServers = null;
+    if (servers.size() > 0) {
+      String sparkIp = System.getenv("SPARK_LOCAL_IP");
+      String hostName = null;
+      try {
+        hostName = sparkIp != null ? InetAddress.getByName(sparkIp).getCanonicalHostName() :
+            InetAddress.getLocalHost().getCanonicalHostName();
+      } catch (Exception e) {
+        hostName = "";
+      }
+      int spacing = servers.size() / 3;
+      if (spacing != 0) {
+        prefServers = new ArrayList<ServerLocation>();
+        ServerLocation localServer = null;
+        for (int j = 0; j < servers.size(); ++j) {
+          ServerLocation sl = servers.get(j);
+          if (localServer == null && sl.getHostName().equals(hostName)) {
+            localServer = sl;
+          } else {
+            if (j + 1 % spacing == 0) {
+              prefServers.add(servers.get(j));
+            }
+          }
+        }
+        if (localServer != null) {
+          if (prefServers.size() < 3) {
+            prefServers.add(0, localServer);
+          } else {
+            prefServers.set(0, localServer);
+          }
+        }
+      } else {
+        // bring local server if any to start position
+        int localServerIndex = -1;
+        for (int j = 0; j < servers.size(); ++j) {
+          ServerLocation sl = servers.get(j);
+          if (localServerIndex == -1 && sl.getHostName().equals(hostName)) {
+            localServerIndex = j;
+            break;
           }
+        }
+        if (localServerIndex != -1) {
+          ServerLocation local = servers.remove(localServerIndex);
+          servers.add(0, local);
+        }
+        prefServers = servers;
```

Comment:
I had understood the code & time of copying , I think the idea is to bring the local host before any other. and since they are sorted on host name, spacing allows to pick up a second host which is different from the local host .